Reduce number of memory copies needed to evaluate long JS strings

bnham opened the pull request:
#### f35ee5f5a5af539cf2343ea3ad9b5b4c3f7dcba4
<pre>
Reduce number of memory copies needed to evaluate long JS strings
<a href="https://bugs.webkit.org/show_bug.cgi?id=304037">https://bugs.webkit.org/show_bug.cgi?id=304037</a>
<a href="https://rdar.apple.com/166338115">rdar://166338115</a>

Reviewed by Alex Christensen.

This has the same goal as 304155@main, which is to reduce the number of string copies associated
with large JS scripts that are repeatedly evaluated in to multiple frames or web views.

This builds on the existing TransferString work (303899@main) and adds a new method
`TransferString::createCached(NSString *)`. This method will either create a new TransferString
instance associated with the NSString (and then cache it using an Obj-C associated object), or reuse
an existing TransferString. By reusing the cached TransferString, we end up reusing the same shared
memory mapping for that string&apos;s bytes across multiple evaluateJavaScript invocations and across
multiple processes.

For now, this optimization is only enabled on the autofill world, since we know this optimization is
profitable in that world. It might not be profitable to do this caching in other worlds (e.g. some
use cases might evaluate lots of large transient script strings). This seems fine for now until we
can develop a better heuristic around when it is profitable to cache the memory mapping.

Finally, reduce `transferAsMappingSize` to a single page so that this optimization applies to more
strings. On the receiving end (in `WebPage::runJavaScript`), the call to `TransferString::release`
uses `transferAsMappingSize` as a hint for whether to reuse the virtual copy or to create a new
physical copy of the bytes. Reduce this to a single page to maximize memory savings.

Test: Tools/TestWebKitAPI/Tests/WebKitCocoa/UserContentController.mm
* Source/WebKit/Platform/IPC/TransferString.h:
* Source/WebKit/Platform/IPC/cocoa/TransferStringCocoa.mm: Added.
(-[_WKTransferStringWrapper initWithString:]):
(-[_WKTransferStringWrapper string]):
(IPC::TransferString::createCached):
* Source/WebKit/Platform/SourcesCocoa.txt:
* Source/WebKit/Platform/cocoa/NetworkIssueReporter.h:
* Source/WebKit/Platform/cocoa/NetworkIssueReporter.mm:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _evaluateJavaScript:asAsyncFunction:withSourceURL:withArguments:forceUserGesture:inFrame:inWorld:completionHandler:]):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/UserContentController.mm:
(TEST(WKUserContentController, EvaluateLargeJavaScriptStringInAutoFillWorld)):

Canonical link: <a href="https://commits.webkit.org/304383@main">https://commits.webkit.org/304383@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/11459ccca005fc4976e2dfe861dab5317de874da

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/135402 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/7781 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/46680 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/143093 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/87114 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/30ef35fe-8227-4eb4-abbe-a7188b1ad22c) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/137271 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/8417 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/7628 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/103477 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/70861 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/2c8c4b5c-fc46-4a03-b1ea-f2f8c3afd753) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/138348 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/6032 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/121357 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/84344 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/1563f8d0-149a-43eb-9e60-ed2ff9963797) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/5815 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/3422 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/3705 "Built successfully") | | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/115015 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/39542 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/145849 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/7464 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/40113 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/111848 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/7505 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/6257 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/112219 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28473 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/5664 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/117658 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/61371 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/7518 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/35781 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/7266 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/71067 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/7485 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/7367 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

kkinnunen-apple commented on Source/WebKit/Platform/IPC/cocoa/TransferStringCocoa.mm:
```diff
@@ -0,0 +1,109 @@
+
+    RetainPtr wrapper = (_WKTransferStringWrapper *)objc_getAssociatedObject(string, transferStringWrapperKey);
+    if (!wrapper) {
+        RefPtr<WebCore::SharedMemory> buffer;
```
Looks good if you think it's good.
I would not perhaps propose this myself, on the account of hanging big memory objects to already big memory objects, causing subtle performance differences that are impossible for the developer to know. OTOH these are academic concerns today, until they are not.

If I may paint the bikeshed:
- Avoid holding the SharedMemory, since that means that the memory is mapped. Instead, hold just the handles. This allows (future) codepaths that create handles without mapping (similar to what TransferString.cpp does).
- Avoid duplicating the string copying code. It already exists in TransferString.cpp.
- Avoid degenerating types ("span16 -> blob + bool==false", "span8 -> blob + bool=true"
- I'd call this just "createCached".

So something along the lines of:
```
@implementation _WKTransferStringWrapper

- (instancetype)initWithSpan8:(TransferString::SharedSpan8&&) handle { .. }
- (instancetype)initWithSpan16:(TransferString::SharedSpan8&&) handle {.. }
- std::optional<TransferString> createTransferString { .. }

@end

std::optional<TransferString> TransferString::createCached(NSString *string)
{
    if (RetainPtr wrapper = (_WKTransferStringWrapper *)objc_getAssociatedObject(string, transferStringWrapperKey))
        return [wrapper createTransferString];
    auto result = TransferString::create(string);
    if (result) {
        WTF::switchOn(result->m_storage,
            [](const String& string) {
            },
            [](const RetainPtr<CFStringRef>& string) {
            },
            [&string](const SharedSpan8& span8) {
                RetainPtr wrapper = adoptNS([[_WKTransferStringWrapper alloc] initWithSpan8: SharedSpan8 { span8 }]);
                objc_setAssociatedObject(string, transferStringWrapperKey, wrapper.get(), OBJC_ASSOCIATION_RETAIN_NONATOMIC);
            },
            [&string](const SharedSpan16& span16) {
                RetainPtr wrapper = adoptNS([[_WKTransferStringWrapper alloc] initWithSpan16: SharedSpan16 { span16 }]);
                objc_setAssociatedObject(string, transferStringWrapperKey, wrapper.get(), OBJC_ASSOCIATION_RETAIN_NONATOMIC);
            });
    }
    return result;
}

```

However, this is just an idea. By no means it's a request to rewrite the already working patch.

bnham replied:
> on the account of hanging big memory objects to already big memory objects, causing subtle performance differences that are impossible for the developer to know.

Definitely a concern, and that's why I limited to the autofill world for now, at least until I can come up with a better heuristic.

> Avoid duplicating the string copying code

I looked this over and it seems like it shouldn't be hard to delete all the duplicated code, which I will do in the upcoming revision.

bnham committed:
Reduce number of memory copies needed to evaluate long JS strings
https://bugs.webkit.org/show_bug.cgi?id=304037
rdar://166338115

Reviewed by Alex Christensen.

This has the same goal as 304155@main, which is to reduce the number of string copies associated
with large JS scripts that are repeatedly evaluated in to multiple frames or web views.

This builds on the existing TransferString work (303899@main) and adds a new method
`TransferString::createCached(NSString *)`. This method will either create a new TransferString
instance associated with the NSString (and then cache it using an Obj-C associated object), or reuse
an existing TransferString. By reusing the cached TransferString, we end up reusing the same shared
memory mapping for that string's bytes across multiple evaluateJavaScript invocations and across
multiple processes.

For now, this optimization is only enabled on the autofill world, since we know this optimization is
profitable in that world. It might not be profitable to do this caching in other worlds (e.g. some
use cases might evaluate lots of large transient script strings). This seems fine for now until we
can develop a better heuristic around when it is profitable to cache the memory mapping.

Finally, reduce `transferAsMappingSize` to a single page so that this optimization applies to more
strings. On the receiving end (in `WebPage::runJavaScript`), the call to `TransferString::release`
uses `transferAsMappingSize` as a hint for whether to reuse the virtual copy or to create a new
physical copy of the bytes. Reduce this to a single page to maximize memory savings.

Test: Tools/TestWebKitAPI/Tests/WebKitCocoa/UserContentController.mm
* Source/WebKit/Platform/IPC/TransferString.h:
* Source/WebKit/Platform/IPC/cocoa/TransferStringCocoa.mm: Added.
(-[_WKTransferStringWrapper initWithString:]):
(-[_WKTransferStringWrapper string]):
(IPC::TransferString::createCached):
* Source/WebKit/Platform/SourcesCocoa.txt:
* Source/WebKit/Platform/cocoa/NetworkIssueReporter.h:
* Source/WebKit/Platform/cocoa/NetworkIssueReporter.mm:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _evaluateJavaScript:asAsyncFunction:withSourceURL:withArguments:forceUserGesture:inFrame:inWorld:completionHandler:]):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/UserContentController.mm:
(TEST(WKUserContentController, EvaluateLargeJavaScriptStringInAutoFillWorld)):

Canonical link: https://commits.webkit.org/304383@main
